### ISSUE TITLE
Add version numbers to generated script files

### ISSUE THREAD

**Issue by fortes** [Treesaver/treesaver#98]
Need to start tracking the version number

**Comment by bramstein** [comment]
Are you planning on tagging each release?

**Comment by bramstein** [comment]
With tags we could do something like this: http://www.hermanradtke.com/blog/canonical-version-numbers-with-git/

**Comment by fortes** [comment]
That's a pretty neat git trick. I don't like it for a filename per se, but I think it'd be useful to have as a variable in the JS (for error reporting?)

So we could do something like:

```
window.TS_VERSION = '2.1pre5-4-g675eae1'
```

So we have an easy way of getting to the right commit.

However, I think it's best to keep our filenames simple, something like:

```
treesaver-0.6.js
treesaver-ui-0.6.js
```

Or whatever.

**Comment by bramstein** [comment]
I had a look at this and there are several options for doing this. The main question is whether or not we want version numbers in the development "build" or not. Adding the version number to the production build (both as a variable and as part of the filename) is quite simple (some changes to `pavement.py`.) Doing it for development as well is more involved and probably requires a local commit hook. I can't think of any need for development version numbers, so I'll leave that for now.

One thing to note is that it will only work with annotated or signed tags, lightweight tags won't work. So tags should be created with `git tag -a 0.2.1`.

We should probably create an initial tag so the system starts working. I quite like semantic versioning (major.minor.patch), should we use that?

**Comment by fortes** [comment]
Agree that there is no need for version numbers in the dev builds -- this is really for the output of the `paver compile` family (I can't think of a reason it'd even be necessary for `paver debug`).

I can't say I've used `git tag` much -- we'll have to use real signing here? Does that require much setup?

I guess the alternative would be to add a version number within the `modules.json` file (or something similar) for the filename. Having the commit info could still be generated without tags if we wished, right?

Semantic versioning works for me. Let's aim for 1.0 being Nomad release.

**Comment by bramstein** [comment]
The only reason I could think of is running a development build on a test environment, but it's not something I do very often. Let's keep it simple for now.

We don't have to use signing, creating an annotated tag also works fine, e.g. `git tag -a 0.9.1` and entering a message.

It is very simple to both get the latest tag (e.g. 0.9.1) and the version number (e.g. 0.9.1-32-ge25a6c) so we could just use the tag name for the filenames and the version number for the `window.TS_VERSION` variable. I haven't done it yet, but I think both can be inserted by the build script without modifying the source.

**Comment by fortes** [comment]
Cool, sounds good.

**Comment by bramstein** [comment]
Turned out to be too simple. I had a go at this, but ran into a bunch of problems. The main problem is that the Closure compiler requires module names to be valid javascript identifiers. This rules out any files with `.` or `-` in it generated by the compiler.

Another problem is that the Treesaver code refers to `treesaver.js` (`scriptloader.js`) explicitly. Changing file names will break this.

If we want to continue down this path, I think the best bet is to do some renames and replacement operations on the generated files (I think we can do this without breaking anything.)

Having the version number as a generated variable is not a problem.

**Comment by fortes** [comment]
Ah yes, we check for the filename explicitly when loading the `ui.js` file (which is also hardcoded).

Some ideas:
1. Instead of hardcoding to `"treesaver.js"` in `scriptloader.js` use a variable name that can be supplied in the build process (too complicated?)
2. Match treesaver.js filename via simple regex, i.e. `/treesaver-(\d+\.\d+)\.js/` or whatever -- will still need to know the filename for the UI module.

Any other ideas?

BTW, feel free to change the way the modules are built if need be. `Modules.json` was an original hacky solution (Closure compiler doesn't have a good module build system yet, so everyone hacks it).

**Comment by bramstein** [comment]
I think I'm still in favour of renaming the bundled files and doing some search and replace in them. It is ugly, but as far as I know, we won't be able to get the closure compiler to generate bundles for us with version numbers in the filenames, so we might as well keep it conceptually simple and put all the magic in the build script.

**Comment by bramstein** [comment]
In the mean time I added the build number as TS_VERSION on the window object, and pushed (a rather arbitrary) 0.9.0 tag so it functions as expected. TS_VERSION returns `dev` on development builds.

**Comment by fortes** [comment]
OK, works for me.

Which do you prefer:

`window.TS_VERSION` or `treesaver.VERSION` -- I'm thinking it might be best to define on the `treesaver` object instead of placing something on the global `window`.

**Comment by bramstein** [comment]
I prefer treesaver.VERSION as well. I'll change it.

**Comment by fortes** [comment]
Cool. Use `goog.exportProperty` or `goog.exportSymbol` for consistency (I forget which right now)

**Comment by bramstein** [comment]
Fixed per SHA: 16dc8ed07cced0d4f99ec72ab72213071600f42a

I came up with an idea on how to improve the module system and bundling, but since it wasn't specifically part of this issue, and a larger task, I created a new issue for it.
